File: DailyChallenges/CPP/LeetCode_953.cpp
class Solution {
public:
    bool isAlienSorted(vector<string>& words, string order) {
        int ord[26];
        for (int i=0; i<26; ++i) ord[order[i]-'a'] = i;
        
        function<bool(const string&, const string&)> cmp = [&](const string& w1, const string& w2){
            if (w1 == w2) return true;
            if (w1.size() > w2.size()) return !cmp(w2, w1);
            for (int i = 0; i< w1.size(); ++i){
                if (w1[i] == w2[i]) continue;
                return ord[w1[i]-'a'] < ord[w2[i]-'a'];
            }
            return true;
        };
        
        for (int i=1; i<words.size(); ++i){
            if (cmp(words[i-1], words[i])) continue;
            return false;
        }
        return true;
    }
};

File: DailyChallenges/Scala/LeetCode_953.scala
object Solution {
    def isAlienSorted(words: Array[String], order: String): Boolean = {
        val n = words.length
        if (n == 1) return true
        
        val dict = order.zipWithIndex.toMap

        def check(w1: String, w2: String): Boolean = 
            if (w1.isEmpty) true 
            else if (w1.nonEmpty && w2.isEmpty) false
            else {
                val (s1, s2) = (dict(w1.head), dict(w2.head))
                if (s1 == s2) check(w1.tail, w2.tail) else if (s1 < s2) true else false             
            }

        for (i <- Range(1,n); w1 = words(i-1); w2 = words(i))
            if (!check(w1, w2)) return false
        
        true
    }
}
